src/main/java/org/opentripplanner/routing/edgetype/TimetableSnapshot.java

```diff
@@ -264,21 +266,46 @@ public TimetableSnapshot commit(boolean force) {
     }
 
     /**
-     * Clear all data of snapshot 
+     * Clear all data of snapshot for the provided feed id
+     *
+     * @param feedId feed id to clear the snapshop for
      */
-    public void clear() {
+    public void clear(String feedId) {
         if (readOnly) {
             throw new ConcurrentModificationException("This TimetableSnapshot is read-only.");
         }
-        
-        // If this snapshot is not empty, it will be dirty after the clear action 
-        if (!timetables.isEmpty() || !lastAddedTripPattern.isEmpty()) {
+        // Clear all data from snapshot.
+        boolean timetableWasModified = clearTimetable(feedId);
+        boolean lastAddedWasModified = clearLastAddedTripPattern(feedId);
+
+        // If this snapshot was modified, it will be dirty after the clear actions.
+        if (timetableWasModified || lastAddedWasModified) {
             dirty = true;
         }
-        
-        // Clear all data from snapshot
-        timetables.clear();
-        lastAddedTripPattern.clear();
+    }
+
+    /**
+     * Clear timetable for all patterns matching the provided feed id.
+     *
+     * @param feedId feed id to clear out
+     * @return true if the timetable changed as a result of the call
+     */
+    protected boolean clearTimetable(String feedId) {
+        return timetables.keySet().removeAll(timetables.keySet().stream()
```

Comment:
Switched to use removeIf, it's also a bit more readable I think. I haven't measured if one is faster than the other though. Thanks!